torch/csrc/jit/passes/frozen_concat_linear.cpp

```diff
@@ -0,0 +1,186 @@
+#include <c10/util/irange.h>
+#include <torch/csrc/jit/ir/alias_analysis.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/ir/ir_views.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/frozen_concat_linear.h>
+#include <torch/csrc/jit/passes/frozen_conv_folding.h>
+#include <torch/csrc/jit/passes/frozen_graph_optimizations.h>
+#include <torch/csrc/jit/passes/remove_dropout.h>
+#include <torch/csrc/jit/passes/utils/optimization_pass.h>
+#include <torch/csrc/jit/runtime/graph_executor.h>
+#include <torch/csrc/utils/memory.h>
+#include <vector>
+#include "ATen/core/functional.h"
+#include "ATen/core/interned_strings.h"
+
+namespace torch {
+namespace jit {
+namespace {
+
+using Tensor = at::Tensor;
+
+class ConcatLinearLayers : public torch::jit::OptimizationPass {
+  using torch::jit::OptimizationPass::OptimizationPass;
+  void handleBlock(Block* b) override {
+    // We are using an ordered list so that one knows that
+    // one only needs to move arguments forward to be a valid move, not
+    // backwards.
+    std::unordered_map<Value*, std::list<Node*>> tensorUsers;
+
+    for (Node* n : b->nodes()) {
+      // Grouping together all linear layers that use the same Tensor for input
+      if (n->kind() != aten::linear) {
+        continue;
+      }
+      
+      // Punting on support for non-constant params for now.
+      // We really just need to know that the shape of the weights and biases are constant.
+      if(nonConstantParameters(n)){
+        continue;
+      }
+
+      Value* op_value = n->inputs().at(0);
+      auto weight = n->namedInput("weight");
+      auto bias = n->namedInput("bias");
+
+      if (!constant_as<Tensor>(weight).has_value() ||
+          !constant_as<Tensor>(bias).has_value()) {
+        continue;
+      }
+      if (weight->type() == NoneType::get() ||
+          bias->type() == NoneType::get()) {
+        continue;
+      }
+
+      if (tensorUsers.find(op_value) == tensorUsers.cend()) {
+        tensorUsers.insert({op_value, std::list<Node*>()});
+      }
+      tensorUsers.find(op_value)->second.push_back(n);
+    }
+
+    // Check the uses of a tensor to find ones that can be combined
+    for (auto& kv : tensorUsers) {
+      Value* op_value = kv.first;
+      auto& uses = kv.second;
+
+      while (!uses.empty()) {
+        auto it_first = uses.begin();
+        Node* base_node = *it_first;
+        uses.erase(it_first);
+
+        auto weight_type = constant_as<Tensor>(base_node->namedInput("weight"))->dtype();
+        auto bias_type = constant_as<Tensor>(base_node->namedInput("bias"))->dtype();
+        std::vector<Node*> matching_nodes;
```

Comment:
nit: I think `matching_nodes` is a little bit misleading, since it doesn't make a lot of sense to say a node is matched to itself. maybe `compatible_linear_layers` ?